server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java

```diff
@@ -0,0 +1,112 @@
+    private static final TransportVersion INFERRED_VERSION = TransportVersion.V_8_8_0;
+
+    private final MasterServiceTaskQueue<NodeTransportVersionTask> masterService;
+    private final ClusterAdminClient client;
+
+    public TransportVersionsFixupListener(ClusterService service, ClusterAdminClient client) {
+        masterService = service.createTaskQueue("fixup-transport-versions", Priority.LANGUID, new TransportVersionUpdater());
+        this.client = client;
+    }
+
+    private record NodeTransportVersionTask(String nodeId, TransportVersion version) implements ClusterStateTaskListener {
+        @Override
+        public void onFailure(Exception e) {
+            Log.error("Could not apply transport version for node {} to cluster state", nodeId, e);
+        }
+    }
+
+    private static class TransportVersionUpdater implements ClusterStateTaskExecutor<NodeTransportVersionTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodeTransportVersionTask> context) throws Exception {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            for (var c : context.taskContexts()) {
+                var t = c.getTask();
+                // this node's transport version might have been updated already/node has gone away
+                if (Objects.equals(builder.transportVersions().get(t.nodeId()), INFERRED_VERSION)) {
+                    builder.putTransportVersion(t.nodeId(), t.version());
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster() == false) return; // only if we're master
+
+        // if the min node version >= 8.8.0, and the min transport version is inferred,
+        // then refresh all inferred transport versions to their real versions
+        // now that everything should understand cluster state with transport versions
+        if (event.state().nodes().getMinNodeVersion().onOrAfter(Version.V_8_8_0)
+            && event.state().getMinTransportVersion().equals(INFERRED_VERSION)) {
+            // find all the relevant nodes
+            String[] nodes = event.state()
+                .transportVersions()
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue().equals(INFERRED_VERSION))
+                .map(Map.Entry::getKey)
+                .toArray(String[]::new);
+
+            NodesInfoRequest request = new NodesInfoRequest(nodes);
+            request.clear().addMetric(ClusterState.Metric.NODES.toString());
+            client.nodesInfo(request, new ActionListener<>() {
+                @Override
+                public void onResponse(NodesInfoResponse response) {
+                    for (NodeInfo n : response.getNodes()) {
+                        String nodeId = n.getNode().getId();
+                        masterService.submitTask(
```

Comment:
Could we do this as a single bulk task? There will be some batching here, but if the master service is idle then the first `submitTask()` will usually trigger the publication a one-task batch, with the other tasks all waiting for the next publication.

Reply:
Done